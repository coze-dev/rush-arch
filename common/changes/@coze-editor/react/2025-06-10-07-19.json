{
  "changes": [
    {
      "packageName": "@coze-editor/react",
      "comment": "tmp",
      "type": "patch"
    }
  ],
  "packageName": "@coze-editor/react",
  "email": "tecvan.fe@qq.com"
}
