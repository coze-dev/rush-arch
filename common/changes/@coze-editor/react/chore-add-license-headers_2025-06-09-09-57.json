{
  "changes": [
    {
      "packageName": "@coze-editor/react",
      "comment": "add license header by running rush license-header",
      "type": "patch"
    }
  ],
  "packageName": "@coze-editor/react",
  "email": "stream-pipe@users.noreply.github.com"
}
