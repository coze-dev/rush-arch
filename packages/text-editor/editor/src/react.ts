//  Copyright (c) 2025 coze-dev
//  SPDX-License-Identifier: MIT

export * from '@coze-editor/react';
export * from '@coze-editor/react-components';
export * from '@coze-editor/react-hooks';
