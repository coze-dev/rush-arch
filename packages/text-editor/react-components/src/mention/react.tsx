//  Copyright (c) 2025 coze-dev
//  SPDX-License-Identifier: MIT

import { useInjectorEffect, useLatest } from '@coze-editor/react-hooks';

import { type MentionOptions } from './types';
import { hasTrigger } from './is';
import { mention, getCurrentMentionReplaceRange } from './extension';

function Mention(props: MentionOptions) {
  const propsRef = useLatest(props);

  useInjectorEffect(injector => {
    const sharedOptions: Partial<MentionOptions> = {
      search: props.search,
      onOpenChange(...args) {
        if (typeof propsRef.current.onOpenChange === 'function') {
          return propsRef.current.onOpenChange(...args);
        }
      },
      onSearch(...args) {
        if (typeof propsRef.current.onSearch === 'function') {
          return propsRef.current.onSearch(...args);
        }
      },
    };
    return injector.inject([
      mention(
        hasTrigger(props)
          ? {
              ...sharedOptions,
              trigger(tr) {
                if (hasTrigger(propsRef.current)) {
                  return propsRef.current.trigger(tr);
                }
              },
            }
          : {
              ...sharedOptions,
              triggerCharacters: props.triggerCharacters ?? [],
            },
      ),
    ]);
  });

  return null;
}

export { Mention, getCurrentMentionReplaceRange };

export type { MentionOptions };
