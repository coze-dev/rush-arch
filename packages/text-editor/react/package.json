{
  "name": "@coze-editor/react",
  "version": "1.0.0-alpha.45",
  "description": "react",
  "license": "MIT",
  "author": "fengzilong",
  "maintainers": [],
  "sideEffects": [
    "**/*.css",
    "**/*.less",
    "**/*.sass",
    "**/*.scss"
  ],
  "main": "./dist/esm/index.js",
  "module": "./dist/esm/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup",
    "dev": "vite --force",
    "lint": "eslint && tsc --noEmit",
    "prepublishOnly": "pnpm run build"
  },
  "dependencies": {
    "@coze-editor/core": "workspace:*"
  },
  "devDependencies": {
    "@codemirror/state": "^6.4.1",
    "@codemirror/view": "^6.26.1",
    "@coze-arch/ts-config": "workspace:*",
    "@coze-editor/eslint-config": "workspace:*",
    "@types/node": "^22",
    "@types/react": "18.2.37",
    "@types/react-dom": "18.2.15",
    "eslint": "9.14.0",
    "react": "~18.2.0",
    "react-dom": "~18.2.0",
    "tsup": "^8.0.1",
    "typescript": "^5.8.2"
  },
  "peerDependencies": {
    "@codemirror/state": "^6.4.1",
    "@codemirror/view": "^6.26.1",
    "react": "~18.2.0",
    "react-dom": "~18.2.0"
  },
  "packageManager": "pnpm@9.15.0",
  "test:main": "./src/index.ts"
}
