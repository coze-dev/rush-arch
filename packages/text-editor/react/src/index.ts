export { Renderer, type InferRendererProps } from './renderer';

export { createRenderer } from './create-renderer';

export { EditorProvider, useEditor, useInjector } from './provider';

export * from '@coze-editor/core';
