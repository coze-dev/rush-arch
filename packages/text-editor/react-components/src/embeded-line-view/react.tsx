//  Copyright (c) 2025 coze-dev
//  SPDX-License-Identifier: MIT

import { createPortal } from 'react-dom';
import { forwardRef, type ReactNode, useImperativeHandle } from 'react';

import {
  useHTMLElement,
  useInjectorEffect,
  useLatest,
} from '@coze-editor/react-hooks';
import { type BuiltinEditorAPI, useEditor } from '@coze-editor/react';
import { Decoration, EditorView, WidgetType } from '@codemirror/view';
import { MapMode, StateEffect, StateField } from '@codemirror/state';

interface EmbededLineViewProps {
  children?: ReactNode;
}

class DOMWidget extends WidgetType {
  constructor(private options: { dom: HTMLElement }) {
    super();
  }

  toDOM() {
    return this.options.dom;
  }

  eq(other: DOMWidget) {
    return this.options.dom === other.options.dom;
  }
}

interface UpdateOptions {
  visible: boolean;
  // line number, 1-based, defaults to 1
  line?: number;
  // where to put the view, before or after the target line
  side?: EmbededLineViewSide;
}

interface EmbededLineViewRefProps {
  update: (options: UpdateOptions) => void;
}

enum EmbededLineViewSide {
  Before = 'before',
  After = 'after',
}

const updateEffect = StateEffect.define<Required<UpdateOptions>>();

const EmbededLineView = forwardRef<
  EmbededLineViewRefProps,
  EmbededLineViewProps
>(function EmbededLineView({ children }, ref) {
  const dom = useHTMLElement('div');
  const domRef = useLatest(dom);
  const editor = useEditor<BuiltinEditorAPI>();
  const editorRef = useLatest(editor);

  useImperativeHandle(ref, () => ({
    update({ visible, line, side }: UpdateOptions) {
      editorRef.current?.$view.dispatch({
        effects: updateEffect.of({
          visible,
          line: line ?? 1,
          side: side ?? EmbededLineViewSide.Before,
        }),
      });
    },
  }));

  useInjectorEffect(injector => {
    const field = StateField.define({
      create() {
        return {
          position: 0,
          side: EmbededLineViewSide.Before,
          decorations: Decoration.none,
        };
      },
      update(value, tr) {
        const newPosition = tr.changes.mapPos(
          value.position,
          -1,
          MapMode.Simple,
        );

        let newValue = value;
        if (typeof newPosition === 'number' && newPosition !== value.position) {
          const docLine = tr.state.doc.lineAt(newPosition);

          let decoPos = 0;

          if (value.side === EmbededLineViewSide.After) {
            decoPos = docLine.to;
          } else {
            decoPos = docLine.from;
          }

          newValue = {
            ...value,
            position: newPosition,
            decorations: Decoration.set([
              Decoration.widget({
                widget: new DOMWidget({
                  dom: domRef.current,
                }),
                block: true,
                side: value.side === EmbededLineViewSide.After ? 1 : -1,
              }).range(decoPos),
            ]),
          };
        }

        for (const effect of tr.effects) {
          if (effect.is(updateEffect)) {
            const { visible, line, side } = effect.value;

            if (!visible) {
              return {
                position: newValue.position,
                side,
                decorations: Decoration.none,
              };
            }

            // out of range, ignore
            if (line < 1 || line > tr.startState.doc.lines) {
              continue;
            }

            const docLine = tr.startState.doc.line(line);

            let decoPos = 0;

            if (side === EmbededLineViewSide.After) {
              decoPos = docLine.to;
            } else {
              decoPos = docLine.from;
            }

            return {
              position: decoPos,
              side,
              decorations: Decoration.set([
                Decoration.widget({
                  widget: new DOMWidget({
                    dom: domRef.current,
                  }),
                  block: true,
                  side: newValue.side === EmbededLineViewSide.After ? 1 : -1,
                }).range(decoPos),
              ]),
            };
          }
        }

        return newValue;
      },
      provide(f) {
        return EditorView.decorations.compute(
          [f],
          state => state.field(f).decorations,
        );
      },
    });

    return injector.inject([field]);
  }, []);

  return createPortal(children, dom);
});

export { EmbededLineView, EmbededLineViewSide, type EmbededLineViewRefProps };
