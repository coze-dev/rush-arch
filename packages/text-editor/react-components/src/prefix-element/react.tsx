import { createPortal } from 'react-dom';
import { type ReactNode, useLayoutEffect } from 'react';

import { useHTMLElement, useLatest } from '@coze-editor/react-hooks';
import { useInjector } from '@coze-editor/react';
import { Decoration, EditorView, keymap, WidgetType } from '@codemirror/view';

import { clientRectsFor, flattenRect } from './dom';

class PrefixElementWidget extends WidgetType {
  constructor(
    readonly content:
      | string
      | HTMLElement
      | ((view: EditorView) => HTMLElement),
  ) {
    super();
  }

  toDOM(view: EditorView) {
    const wrap = document.createElement('span');
    wrap.className = 'cm-prefix-element';
    // solved the cursor height problem(eg. cursor double height with two lines)
    wrap.style.cssText = 'height: 0;';
    wrap.appendChild(
      typeof this.content === 'string'
        ? document.createTextNode(this.content)
        : typeof this.content === 'function'
          ? this.content(view)
          : this.content,
    );
    if (typeof this.content === 'string') {
      wrap.setAttribute('aria-label', `prefix ${this.content}`);
    } else {
      wrap.setAttribute('aria-hidden', 'true');
    }
    return wrap;
  }

  coordsAt(dom: HTMLElement) {
    const rects = dom.firstChild ? clientRectsFor(dom.firstChild) : [];
    if (!rects.length) {
      return null;
    }
    const style = window.getComputedStyle(dom.parentNode as HTMLElement);
    const rect = flattenRect(rects[0], style.direction != 'rtl');
    const lineHeight = parseInt(style.lineHeight);
    if (rect.bottom - rect.top > lineHeight * 1.5) {
      return {
        left: rect.left,
        right: rect.right,
        top: rect.top,
        bottom: rect.top + lineHeight,
      };
    }
    return rect;
  }

  ignoreEvent() {
    return false;
  }
}

interface PrefixElementProps {
  deletable?: boolean;
  onDelete?: () => void;
  children?: ReactNode;
}

function PrefixElement({ deletable, onDelete, children }: PrefixElementProps) {
  const element = useHTMLElement('span');
  const latestElement = useLatest(element);
  const latestDeletable = useLatest(deletable);
  const latestOnDelete = useLatest(onDelete);
  const injector = useInjector();

  useLayoutEffect(() => {
    function run(view: EditorView) {
      if (
        view.state.selection.main.empty &&
        view.state.selection.main.head === 0 &&
        latestDeletable.current === true &&
        typeof latestOnDelete.current === 'function'
      ) {
        latestOnDelete.current();
      }

      return false;
    }

    return injector.inject([
      EditorView.decorations.of(
        Decoration.set([
          Decoration.widget({
            side: -100,
            block: false,
            widget: new PrefixElementWidget(() => latestElement.current),
          }).range(0),
        ]),
      ),
      keymap.of([
        { key: 'Backspace', shift: run, run },
        { key: 'Meta-Backspace', shift: run, run },
      ]),
    ]);
  }, []);

  return createPortal(children, element);
}

export { PrefixElement };
