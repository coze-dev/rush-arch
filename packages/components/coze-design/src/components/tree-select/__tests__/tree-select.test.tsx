//  Copyright (c) 2025 coze-dev
//  SPDX-License-Identifier: MIT

import '@testing-library/jest-dom';
import React from 'react';

import { fireEvent, render, screen } from '@testing-library/react';

import { TreeSelect } from '..';

const treeData = [
  {
    label: '亚洲',
    value: 'Asia',
    key: '0',
    children: [
      {
        label: '中国',
        value: 'China',
        key: '0-0',
        children: [
          {
            label: '北京',
            value: 'Beijing',
            key: '0-0-0',
          },
          {
            label: '上海',
            value: 'Shanghai',
            key: '0-0-1',
          },
        ],
      },
    ],
  },
  {
    label: '北美洲',
    value: 'North America',
    key: '1',
  },
];

describe('TreeSelect', () => {
  it('should render', () => {
    const { container } = render(
      <TreeSelect
        multiple
        leafOnly={true}
        treeData={treeData}
        placeholder="请选择"
        className="w-200"
      />,
    );
    const element = container.querySelector('.coz-tree-select');
    expect(element).toBeInTheDocument();
  });

  it('should click', async () => {
    const { container } = render(
      <TreeSelect
        leafOnly={true}
        treeData={treeData}
        placeholder="请选择"
        className="w-200"
      />,
    );
    const element = container.querySelector('.coz-tree-select');
    expect(element).toBeInTheDocument();
    // @ts-expect-error -- linter-disable-autofix
    fireEvent.click(element);
    const treeItemShow = await screen.findByRole('tree');
    expect(
      treeItemShow.closest('.coz-tree-select-popover'),
    ).toBeInTheDocument();
    expect(treeItemShow.closest('.coz-tree-select-popover')).toBeVisible();
    const treeNode = screen.queryByText('北美洲');
    expect(treeNode).toBeInTheDocument();
    // @ts-expect-error -- linter-disable-autofix
    fireEvent.click(treeNode.parentNode);
    const selectedNode = container.querySelector('.semi-tree-select-selection');
    expect(selectedNode).toBeInTheDocument();
    expect(selectedNode).toHaveTextContent('北美洲');
  });

  it('should only allow leaf node selection when onlyLeafSelectable is true', async () => {
    const { container } = render(
      <TreeSelect
        treeData={treeData}
        placeholder="请选择"
        onlyLeafSelectable
      />,
    );

    const element = container.querySelector('.coz-tree-select');
    expect(element).toBeInTheDocument();
    // @ts-expect-error -- linter-disable-autofix
    fireEvent.click(element);

    const tree = await screen.findByRole('tree');

    expect(tree.innerHTML).toMatchSnapshot();
  });
});

it('should multiple select', async () => {
  const { container } = render(
    <TreeSelect
      multiple
      leafOnly={true}
      treeData={treeData}
      placeholder="请选择"
      className="w-200"
    />,
  );
  const element = container.querySelector('.coz-tree-select');
  expect(element).toBeInTheDocument();
  // @ts-expect-error -- linter-disable-autofix
  fireEvent.click(element);
  const treeItemShow = await screen.findByRole('tree');
  expect(treeItemShow.closest('.coz-tree-select-popover')).toBeInTheDocument();
  const treeNode = screen.queryByText('亚洲');
  expect(treeNode).toBeInTheDocument();
  // @ts-expect-error -- linter-disable-autofix
  fireEvent.click(treeNode);
  const selectedNode = container.querySelector('.semi-tree-select-selection');
  expect(selectedNode).toBeInTheDocument();
  expect(selectedNode).toHaveTextContent('北京上海');
});
