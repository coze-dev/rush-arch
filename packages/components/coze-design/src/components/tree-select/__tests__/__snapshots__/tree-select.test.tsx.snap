// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TreeSelect > should only allow leaf node selection when onlyLeafSelectable is true 1`] = `"<li class="semi-tree-option semi-tree-option-level-1 semi-tree-option-fullLabel-level-1 semi-tree-option-collapsed" role="treeitem"><span class="semi-tree-option-indent"></span><span class="semi-tree-option-indent"></span><span role="button" aria-label="Collapse the tree item" class="semi-icon semi-icon-small semi-icon-tree_triangle_down semi-tree-option-expand-icon"><svg viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" focusable="false" aria-hidden="true"><path d="m20.56 9.66-7.8 8.97a1 1 0 0 1-1.51 0L3.44 9.66A1 1 0 0 1 4.19 8h15.62a1 1 0 0 1 .75 1.66Z" fill="currentColor"></path></svg></span>亚洲</li><li class="semi-tree-option semi-tree-option-level-1 semi-tree-option-fullLabel-level-1 semi-tree-option-collapsed semi-tree-option-tree-node-last-leaf" role="treeitem"><span class="semi-tree-option-indent"></span><span class="semi-tree-option-empty-icon"></span>北美洲</li>"`;
